code/game/machinery/kitchen/smartfridge.dm

```diff
@@ -128,194 +126,253 @@
 		if(shock(user, 100))
 			return
 
-	ui_interact(user)
+	tgui_interact(user)
 
-/obj/structure/machinery/smartfridge/proc/add_item(var/obj/item/O)
-	O.forceMove(src)
+/obj/structure/machinery/smartfridge/proc/add_local_item(var/obj/item/O)
+	add_item(item_quants, O)
 
-	if(item_quants[O.name])
-		item_quants[O.name]++
+/obj/structure/machinery/smartfridge/proc/add_item(var/list/target, var/obj/item/O)
+	O.forceMove(src)
+	if(target[O.name])
+		target[O.name] += O
 	else
-		item_quants[O.name] = 1
+		target[O.name] = list(O)
 
 /obj/structure/machinery/smartfridge/proc/add_network_item(var/obj/item/O)
 	if(is_in_network())
-		chemical_data.shared_item_storage.Add(O)
-
-		if(chemical_data.shared_item_quantity[O.name])
-			chemical_data.shared_item_quantity[O.name]++
-		else
-			chemical_data.shared_item_quantity[O.name] = 1
+		add_item(chemical_data.shared_item_storage, O)
 		return TRUE
 	return FALSE
 
 //*******************
 //*   SmartFridge Menu
 //********************/
 
-/obj/structure/machinery/smartfridge/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
-	user.set_interaction(src)
-
-	var/data[0]
-	data["contents"] = null
-	data["wires"] = null
-	data["panel_open"] = panel_open
-	data["electrified"] = seconds_electrified > 0
-	data["shoot_inventory"] = shoot_inventory
-	data["locked"] = locked
-	data["secure"] = is_secure_fridge
-	data["networked"] = is_in_network()
-	data["transfer_mode"] = transfer_mode
-
-	var/list/items[0]
-	for (var/i=1 to length(item_quants))
-		var/K = item_quants[i]
-		var/count = item_quants[K]
-		if (count > 0)
-			items.Add(list(list("display_name" = html_encode(capitalize(K)), "vend" = i, "quantity" = count)))
+/obj/structure/machinery/smartfridge/tgui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "SmartFridge", name)
+		ui.open()
 
-	if (length(items) > 0)
-		data["contents"] = items
+/obj/structure/machinery/smartfridge/ui_state(mob/user)
+	return GLOB.not_incapacitated_and_adjacent_strict_state
 
-	if(is_in_network())
-		var/list/networked_items = list()
-		for (var/i=1 to length(chemical_data.shared_item_quantity))
-			var/K = chemical_data.shared_item_quantity[i]
-			var/count = chemical_data.shared_item_quantity[K]
-			if (count > 0)
-				networked_items.Add(list(list("display_name" = html_encode(capitalize(K)), "vend" = i, "quantity" = count)))
+/obj/structure/machinery/smartfridge/ui_status(mob/user, datum/ui_state/state)
+	. = ..()
+	if(inoperable())
+		return UI_CLOSE
+
+/obj/structure/machinery/smartfridge/ui_static_data(mob/user)
+	. = ..(user)
```

Comment:
? why user in the brackets here

also not sure if this needs a parent call at all

Reply:
I think this was me early-optimising. The idea was that the handling of hacking/electrical stuff would be handled upstream. The proc contains user as a param so I thought you would want to pass it in (like most languages) - I'm guessing that isn't needed?

In any case, upstream of this does nothing right now so not going to pre-optimise.